core/src/main/java/org/apache/iceberg/variants/SerializedObject.java

```diff
@@ -0,0 +1,198 @@
+  private final int offsetSize;
+  private final int offsetListOffset;
+  private final int dataOffset;
+  private final VariantValue[] values;
+
+  private SerializedObject(SerializedMetadata metadata, ByteBuffer value, int header) {
+    this.metadata = metadata;
+    this.value = value;
+    this.offsetSize = 1 + ((header & OFFSET_SIZE_MASK) >> OFFSET_SIZE_SHIFT);
+    this.fieldIdSize = 1 + ((header & FIELD_ID_SIZE_MASK) >> FIELD_ID_SIZE_SHIFT);
+    int numElementsSize = ((header & IS_LARGE) == IS_LARGE) ? 4 : 1;
+    int numElements =
+        VariantUtil.readLittleEndianUnsigned(value, Variants.HEADER_SIZE, numElementsSize);
+    this.fieldIdListOffset = Variants.HEADER_SIZE + numElementsSize;
+    this.fieldIds = new Integer[numElements];
+    this.offsetListOffset = fieldIdListOffset + (numElements * fieldIdSize);
+    this.dataOffset = offsetListOffset + ((1 + numElements) * offsetSize);
+    this.values = new VariantValue[numElements];
+  }
+
+  @VisibleForTesting
+  int numElements() {
+    return fieldIds.length;
+  }
+
+  SerializedMetadata metadata() {
+    return metadata;
+  }
+
+  Iterable<Pair<String, Integer>> fields() {
+    return () ->
+        new Iterator<>() {
+          private int index = 0;
+
+          @Override
+          public boolean hasNext() {
+            return index < fieldIds.length;
+          }
+
+          @Override
+          public Pair<String, Integer> next() {
+            Pair<String, Integer> next = Pair.of(metadata.get(id(index)), index);
+            index += 1;
+            return next;
+          }
+        };
+  }
+
+  public Iterable<String> fieldNames() {
+    return () ->
+        new Iterator<>() {
+          private int index = 0;
+
+          @Override
+          public boolean hasNext() {
+            return index < fieldIds.length;
+          }
+
+          @Override
+          public String next() {
+            int id = id(index);
+            index += 1;
+            return metadata.get(id);
+          }
+        };
+  }
+
+  private int id(int index) {
+    if (null == fieldIds[index]) {
+      fieldIds[index] =
+          VariantUtil.readLittleEndianUnsigned(
+              value, fieldIdListOffset + (index * fieldIdSize), fieldIdSize);
+    }
+    return fieldIds[index];
+  }
+
+  @Override
+  public VariantValue get(String name) {
+    // keys are ordered lexicographically by the name
+    int index = VariantUtil.find(fieldIds.length, name, pos -> metadata.get(id(pos)));
+
+    if (index < 0) {
+      return null;
+    }
+
+    if (null == values[index]) {
+      int offset =
+          VariantUtil.readLittleEndianUnsigned(
+              value, offsetListOffset + (index * offsetSize), offsetSize);
+      int next =
+          VariantUtil.readLittleEndianUnsigned(
+              value, offsetListOffset + ((1 + index) * offsetSize), offsetSize);
+      values[index] =
+          Variants.from(metadata, VariantUtil.slice(value, dataOffset + offset, next - offset));
```

Comment:
There is an inconsistency between writer and reader here. Here we are assuming that offsets are in order but the writer in Spark would have different order.

For an object, since we access and write the values 1234, "joe",  and "smith" in order, the offsets will be 3, 0, 7, 13. With that, if we slice the buffer by (0, 7), we get the wrong result. 
{
  "id": 1234,
  "first_name": "joe",
  "last_name": "smith"
}

Reply:
Good catch, I think you're right. I'll update this and add a test.

Reply:
I think the problem is not the (0,7) slice, but the (3,0) slice. That one would have a negative length that would cause a failure. Extra bytes in the buffer are not ideal, but because the length of the buffer isn't used to determine the end of a value it would be safe to include extra bytes.

Reply:
That's right. It will fail at (3,0) slice.

Reply:
I have a test for this and it's passing. Thanks for catching this, @aihuaxu!